### PR TITLE
fix: STAC files should comply to 1.0.0-beta.2 of the specification

### PR DESCRIPTION
**Pull request by jacott**
* Add linz basemaps schema.json
* Fix `extent/spatial/bbox` to `number[][]`
* `description` is always required
* Set `stac_version` to `1.0.0-beta.2`
* Don't use `projection` extension for `collection.json`

### CODE REVIEW

File: packages/cli/src/cog/cog.stac.job.ts [hunk 1]
```diff
@@ -191,9 +196,11 @@ export class CogStacJob implements CogJob {
 
         const sourceProj = job.sourcePtms.proj;
 
-        const bbox = sourceProj.boundsToWgs84BoundingBox(
-            metadata.bounds.map((a) => Bounds.fromJson(a)).reduce((sum, a) => sum.union(a)),
-        );
+        const bbox = [
+            sourceProj.boundsToWgs84BoundingBox(
```
Comment (blacha) [inline packages/cli/src/cog/cog.stac.job.ts:200]: Does this break with the antimeridian?

is the reason their are multiple bbox's for the antimeridian?

File: packages/landing/static/json-schema/basemaps-extension/1.0/schema.json [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        {"$ref": "https://schemas.stacspec.org/v1.0.0-beta.2/item-spec/json-schema/item.json"},
+        {"$ref": "#/definitions/linz-common"},
+        {"$ref": "#/definitions/linz-item"}
+      ],
+    },
+    {
+      "allOf": [
+        {"$ref": "https://schemas.stacspec.org/v1.0.0-beta.2/collection-spec/json-schema/collection.json"},
+        {"$ref": "#/definitions/linz-common"},
+        {"$ref": "#/definitions/linz-collection"},
```
Comment (blacha) [inline packages/landing/static/json-schema/basemaps-extension/1.0/schema.json:18]: we cant use `linz-common` maybe `linz-basemaps-common` ?

File: packages/shared/src/stac/index.ts [hunk 3]
```diff
@@ -1,5 +1,6 @@
-export const StacVersion = '1.0.0';
+export const StacVersion = '1.0.0-beta.2';
 export const StacLicense = 'CC-BY-4.0';
+export const StacBaseMapsExtension = 'https://basemaps.linz.govt.nz/json-schema/basemaps-extension/1.0/schema.json';
```
Comment (blacha) [inline packages/shared/src/stac/index.ts:3]: is this URL a standard for this schema?

How is the file deployed?

File: packages/cli/src/cog/cog.stac.job.ts (hunk 1 above)
Comment (jacott) [reply packages/cli/src/cog/cog.stac.job.ts:200]: No. only one bbox is needed event when AM is crossed; it just means the first longitude will be greater than the second longitude

File: packages/landing/static/json-schema/basemaps-extension/1.0/schema.json (hunk 2 above)
Comment (jacott) [reply packages/landing/static/json-schema/basemaps-extension/1.0/schema.json:18]: It is just a local ref; we can use whatever we want

File: packages/shared/src/stac/index.ts (hunk 3 above)
Comment (jacott) [reply packages/shared/src/stac/index.ts:3]: There is no standard for vendor extensions. I am trying to use a somewhat canonical path.

The file should be deployed as part of the landing package (assuming landing deploys everything under `static`)

File: packages/landing/static/json-schema/basemaps-extension/1.0/schema.json (hunk 2 above)
Comment (jacott) [reply packages/landing/static/json-schema/basemaps-extension/1.0/schema.json]: I changed it to just "common", "collection" and "item"

File: packages/shared/src/stac/index.ts (hunk 3 above)
Comment (jacott) [reply packages/shared/src/stac/index.ts]: I see that the deploy does not deploy everything under static. How is `favicon-16x16.png` etc getting deployed?

Comment (blacha) [reply packages/shared/src/stac/index.ts]: >  I see that the deploy does not deploy everything under static. How is favicon-16x16.png etc getting deployed?

Not too sure maybe I broke it a while back? possibly should be fixed :)

Comment (jacott) [reply packages/shared/src/stac/index.ts]: It is deploying every file under dist but not descending directories; now it does.

Comment (jacott) [reply packages/shared/src/stac/index.ts]: Actually only deploy copies everything under `dist`. Bundle needs to be modified to copy everything from `static` to `dist`

Comment (jacott) [reply packages/shared/src/stac/index.ts]: `bundle` now recursively copies `static` files (not `index.css` etc) to `dist`

File: packages/shared/src/stac/index.ts [hunk 4]
```diff
@@ -1,5 +1,7 @@
-export const StacVersion = '1.0.0';
+export const StacVersion = '1.0.0-beta.2';
 export const StacLicense = 'CC-BY-4.0';
+export const StacBaseMapsExtension =
```
Comment (blacha) [inline packages/shared/src/stac/index.ts:3]: Do we hard code `basemaps.linz.govt.nz` anywhere else? Ideally we should be injecting this later?

Could we make it a config variable inside the bundler?

Comment (jacott) [reply packages/shared/src/stac/index.ts:3]: Do we know the host at run time for `cli `and `bathymetry` packaes and at deploy time for `landing`?

File: packages/landing/static/json-schema/stac-basemaps-extension/1.0/schema.json [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+          "$ref": "#/definitions/collection"
+        }
+      ]
+    }
+  ],
+  "definitions": {
+    "sampling-method": {
+      "type": "string",
+      "enum": [
+        "nearest",
```
Comment (blacha) [inline packages/landing/static/json-schema/stac-basemaps-extension/1.0/schema.json:38]: do you have any thoughts on how we can keep this upto date?

I wonder if we should just use a typescript -> jsonschema to create this.

File: packages/landing/scripts/bundle.js [hunk 6]
```diff
@@ -99,8 +105,31 @@ async function bundleHtml(opts) {
...
+            await fs.mkdir(destPath);
+            await recurseStatic(filePath);
+            continue;
+        }
+        await fs.writeFile(destPath, await fs.readFile(srcPath));
+    }
+}
+
 async function main() {
+    await fs.rmdir('dist', { recursive: true });
```
Comment (blacha) [inline packages/landing/scripts/bundle.js:130]: I dont think we want to delete this here.

I believe there is `clean` if you want to clean it.

File: packages/landing/scripts/bundle.js [hunk 7]
```diff
@@ -99,8 +105,31 @@ async function bundleHtml(opts) {
     await fs.writeFile('./dist/index.html', htmlOutput);
 }
 
+async function recurseStatic(subDir = '') {
```
Comment (blacha) [inline packages/landing/scripts/bundle.js:108]: Could we get a comment as to what this does?

This also seems to be quite similar to the logic below could they use a shared function for the work?

File: packages/landing/scripts/bundle.js [hunk 8]
```diff
@@ -99,8 +105,31 @@ async function bundleHtml(opts) {
...
+    const srcDir = './static' + subDir;
+    const files = await fs.readdir(srcDir);
+    for (const fileName of files) {
+        const filePath = `${subDir}/${fileName}`;
+
+        console.log(`DEBUG SPECIAL_FILES[filePath]`, SPECIAL_FILES[filePath], filePath);
```
Comment (blacha) [inline packages/landing/scripts/bundle.js:114]: do we need this console.log?

File: packages/cli/src/cog/cog.stac.job.ts [hunk 9]
```diff
@@ -150,7 +151,17 @@ export class CogStacJob implements CogJob {
             description = sourceStac.description;
             interval.push(...(sourceStac.extent?.temporal?.interval ?? []));
             links.push({ href: sourceCollectionPath, rel: 'sourceImagery', type: 'application/json' });
-            if (sourceStac.providers != null) providers.push(...sourceStac.providers);
+            if (sourceStac.providers != null) {
+                for (const p of sourceStac.providers) {
+                    if (p.roles.indexOf('host') == -1) {
+                        if (p.url === 'unknown') {
+                            // LINZ LDS has put unknown in some urls
+                            p.url = undefined;
```
Comment (blacha) [inline packages/cli/src/cog/cog.stac.job.ts:159]: is unknown not allowed in the spec?

File: packages/shared/src/stac/index.ts (hunk 4 above)
Comment (jacott) [reply packages/shared/src/stac/index.ts:3]: I think, because this is a link to a LINZ specific schema, the well known location should always be used.

File: packages/landing/static/json-schema/stac-basemaps-extension/1.0/schema.json (hunk 5 above)
Comment (jacott) [reply packages/landing/static/json-schema/stac-basemaps-extension/1.0/schema.json:38]: I don't think the list should change unless the schema version also changes.

File: packages/landing/scripts/bundle.js (hunk 6 above)
Comment (jacott) [reply packages/landing/scripts/bundle.js:130]: I think a bundle should clear the directory because there could be leftover files there that are no longer in `static`. I'm okay with not though if you prefer not to.

File: packages/landing/scripts/bundle.js (hunk 8 above)
Comment (jacott) [reply packages/landing/scripts/bundle.js:114]: oops debug leftover

File: packages/cli/src/cog/cog.stac.job.ts (hunk 9 above)
Comment (jacott) [reply packages/cli/src/cog/cog.stac.job.ts:159]: correct

File: packages/cli/src/cog/cog.stac.job.ts (hunk 1 above)
Comment (palmerj) [reply packages/cli/src/cog/cog.stac.job.ts:210]: Hi Guys, it's best practise to split across the anti-meridian for interoperability reasons with clients. See https://tools.ietf.org/html/rfc7946#section-3.1.9

File: packages/shared/src/stac/index.ts (hunk 4 above)
Comment (blacha) [reply packages/shared/src/stac/index.ts:3]: > because this is a link to a LINZ specific schema

Ok good reasoning!

File: packages/landing/static/json-schema/stac-basemaps-extension/1.0/schema.json (hunk 5 above)
Comment (blacha) [reply packages/landing/static/json-schema/stac-basemaps-extension/1.0/schema.json:38]: If I go add another sampling-method to the code and it generates the stac files from that code, this schema is not representative of what is actually being built.

Is this caught anywhere? Like do I know that adding another sampling method to the source will break this schema?

File: packages/landing/scripts/bundle.js (hunk 6 above)
Comment (blacha) [reply packages/landing/scripts/bundle.js:133]: All the other bundlers I use, webpack, parcel etc, have a very specific "clean" function if things are being deleted.

The static folder in theory is only deployed from a clean state so it shouldnt normally matter, and if you are doing a deployment you really want to `clean` the entire repo.

TLDR: yes please take it out.

File: packages/cli/src/cog/cog.stac.job.ts (hunk 1 above)
Comment (blacha) [reply packages/cli/src/cog/cog.stac.job.ts:210]: @jacott with Jeremy's comment maybe we should split it on the antimeridian? Does the stac spec tell us not to?

File: packages/landing/scripts/bundle.js (hunk 7 above)
Comment (jacott) [reply packages/landing/scripts/bundle.js:112]: DRYing up this code will involve more insertions than deletions but okay.

File: packages/landing/static/json-schema/stac-basemaps-extension/1.0/schema.json (hunk 5 above)
Comment (jacott) [reply packages/landing/static/json-schema/stac-basemaps-extension/1.0/schema.json:38]: I'll make the schema accept any string

File: packages/cli/src/cog/cog.stac.job.ts (hunk 1 above)
Comment (jacott) [reply packages/cli/src/cog/cog.stac.job.ts:210]: @blacha , @palmerj The GeoJSON RFC does not split `bbox`. STAC uses the GeoJSON for items but I defer to others and will split if you think I should.

Comment (jacott) [reply packages/cli/src/cog/cog.stac.job.ts:210]: What the link above is for is polygons which we do split.

Comment (palmerj) [reply packages/cli/src/cog/cog.stac.job.ts:210]: Ok thanks. You are correct. As long as we split the footprint geometry then we are ok. Clients querying on the bbox via the STAC API will of course need to normalise their input bboxes

File: packages/landing/scripts/deploy.js [hunk 10]
```diff
@@ -40,36 +41,45 @@ async function deploy() {
...
-        // Only upload files if they have changed
-        if (oldHash == hash) {
-            console.log('Skipped', fileName, `${(fileData.byteLength / 1024).toFixed(2)}Kb`, hash);
-            continue;
-        }
-        hasChanges = true;
+    /**
+     * Upload all files from the dist directory, including any sub directories, to S3 bucket .
+     */
+    const uploadDirectory = async () => {
```
Comment (blacha) [inline packages/landing/scripts/deploy.js:48]: NIT: Why make a function and call it directly after? could you not just execute the function

**Review by blacha — APPROVED** [review]
